tensorflow/core/kernels/fft_ops.cc

```diff
@@ -97,9 +84,141 @@ class FFTGPUBase : public OpKernel {
+  bool IsForward() const override { return Forward; }
+  bool IsReal() const override { return _Real; }
+
+  void DoFFT(OpKernelContext* ctx, const Tensor& in, uint64* fft_shape,
+             Tensor* out) override {
+    if (!IsReal()) {
+      auto input = ((Tensor) in).flat_inner_dims<complex64, FFTRank + 1>();
+
+      // Apply the functor.
+      FFTFunctor<CPUDevice, complex64, complex64, Eigen::BothParts,
+                Forward ? Eigen::FFT_FORWARD : Eigen::FFT_REVERSE,
+                FFTRank> functor;
+      functor(ctx->eigen_device<CPUDevice>(),
+              out->flat_inner_dims<complex64, FFTRank + 1>(), input);
+    }
+    else {
+      if (IsForward()) {
+        auto input = ((Tensor) in).flat_inner_dims<float, FFTRank + 1>();
+        // Create a temporary placeholder for the full FFT.
+        Tensor temp;
+        OP_REQUIRES_OK(ctx,ctx->allocate_temp(
+          DataTypeToEnum<complex64>::v(), in.shape(), &temp
+        ));
+        auto full_fft = temp.flat_inner_dims<complex64, FFTRank + 1>();
+        // Apply the functor.
+        FFTFunctor<CPUDevice, float, complex64, Eigen::BothParts,
+                  Eigen::FFT_FORWARD, FFTRank> functor;
+        functor(ctx->eigen_device<CPUDevice>(), full_fft, input);
+        // Create zero indices for slicing.
+        Eigen::DSizes<Eigen::DenseIndex, FFTRank + 1> startIndices;
+        // Convert output to tensor and get tensor size for slicing.
+        auto output = out->flat_inner_dims<complex64, FFTRank + 1>();
+        auto sizes = output.dimensions();
+        // Slice the full FFT to get the non-negative frequency components only.
+        output.slice(startIndices, sizes) =
+          full_fft.slice(startIndices, sizes);
+      }
+      else {
+        auto input = ((Tensor) in).flat_inner_dims<complex64, FFTRank + 1>();
+        // The first dimension contains the zero-frequency component which we
+        // do not want to duplicate. So we reconstruct the complex signal by
+        // (1) slicing from the second element, (2) reversing the order,
+        // (3) taking the complex conjugate, (4) concatenating with the original
+        // input. Note that for an even input length, the last element is the
+        // Nyquist frequency which we also do not want to duplicate.
+        Eigen::DSizes<Eigen::DenseIndex, FFTRank + 1> startIndices;
+        startIndices[FFTRank] = 1;
+        auto sizes = input.dimensions();
+        if (sizes[FFTRank] % 2 == 0) {
+          sizes[FFTRank] -= 1;
+        }
+        auto cc = input.slice(startIndices, sizes).conjugate()
+          .reverse(FFTRank);
+        auto full_fft = input.concatenate(cc, FFTRank);
+
+        // Evaluate the IFFT
+        auto output = out->flat_inner_dims<float, FFTRank + 1>();
+        FFTFunctor<CPUDevice, complex64, float, Eigen::RealPart,
+                  Eigen::FFT_REVERSE, FFTRank> functor;
+        functor(ctx->eigen_device<CPUDevice>(), output, full_fft);
```

Comment:
Great, thanks for the insight. I think I tried casting without any luck before but might give it another go. 

Another option could be to allocate a temporary tensor, do all the preprocessing, assign to the temporary tensor to evaluate, compute the IFFT of the temporary tensor. It would lose some of the elegance of Eigen but might be an acceptable workaround.